In [1]:
from bs4 import BeautifulSoup
import requests
import time, os

C:\Users\MERVE\anaconda3\lib\site-packages\requests\__init__.py:89: RequestsDependencyWarning: urllib3 (1.26.12) or chardet (3.0.4) doesn't match a supported version!
  warnings.warn("urllib3 ({}) or chardet ({}) doesn't match a supported "


In [2]:
import pandas as pd
import numpy as np

In [3]:
from selenium import webdriver
from selenium.webdriver.common.keys import Keys

chromedriver = "C:/Users/MERVE/Desktop/chromedriver" # path to the chromedriver executable
os.environ["webdriver.chrome.driver"] = chromedriver

In [4]:
query = "laptop"  # trendyolda laptop aratacağım.
trendyol_search = f"https://www.trendyol.com/sr?q={query}&qt={query}&st={query}&os=1"
url_list = [trendyol_search+"&pi="+str(page) for page in range(1,209)] # ilk 208 sayfadaki bilgisayarları alıyorum.

In [5]:
computers = []
for url in url_list:
    driver = webdriver.Chrome(chromedriver)
    driver.get(url)
    soup_main = BeautifulSoup(driver.page_source, 'html.parser')
    #her sayfada 24 adet laptop var. 
    links = ["https://www.trendyol.com/"+link.a.get("href") for link in soup_main.find_all("div",{"class":"p-card-chldrn-cntnr card-border"})]
    for link in links:
        driver.get(link)
        soup = BeautifulSoup(driver.page_source, 'html.parser')
        brand = soup.find("h1",{"class":"pr-new-br"}).text
        title = soup.find("h1",{"class":"pr-new-br"}).span.text
        try:
            stars = 0
            for star in soup.find("div",{"class":"ratings readonly"}).find_all("div",{"class":"star-w"}):
                stars += int(star.div.findNext().findNext().get("style")[7:10].replace("%","").replace("px",""))            
        except:
            stars = 0
        stars /= 100 
        price = soup.find("span",{"class":"prc-dsc"}).text.split(" ")[0]
        attributes = [(attribute.span.text,attribute.span.findNext().text) for attribute in soup.find("ul",{"class":"detail-attr-container"}).find_all("li",{"class":"detail-attr-item"})]
        attributes = [("Brand",brand), ("Title",title)]+ attributes+[("Star",stars),("Price",price)]
        computers.append(attributes)
driver.quit()

<ipython-input-5-ade393423c4c>:3: DeprecationWarning: executable_path has been deprecated, please pass in a Service object
  driver = webdriver.Chrome(chromedriver)


AttributeError: 'NoneType' object has no attribute 'text'

In [10]:
url

'https://www.trendyol.com/sr?q=laptop&qt=laptop&st=laptop&os=1&pi=206'

In [11]:
len(computers)

4934

## İlk 206 sayfadaki veriyi çektim. Ama bundan sonra Trendyol filtreleme ile arama yapın uyarısı verdiği için error aldım. Yine de üzerinde çalışacabileceğim 4934 tane laptop verisini çekebildim. 

In [12]:
columns= []
for attribute in computers:
    for attribute_name in attribute:
        if attribute_name[0] not in columns:
            columns.append(attribute_name[0])
columns.sort()

In [13]:
columns

['Arttırılabilir Azami Bellek',
 'Bağlantılar',
 'Brand',
 'Cihaz Ağırlığı',
 'Dokunmatik Ekran',
 'Ekran Boyutu',
 'Ekran Kartı',
 'Ekran Kartı Bellek Tipi',
 'Ekran Kartı Hafızası',
 'Ekran Kartı Tipi',
 'Ekran Yenileme Hızı',
 'Ekran Özelliği',
 'Garanti Süresi',
 'Garanti Tipi',
 'Görüntü Kalitesi',
 'HDMI',
 'Hard Disk Kapasitesi',
 'Kapasite',
 'Klavye',
 'Kullanım Amacı',
 'Maksimum İşlemci Hızı (GHz)',
 'Okula Dönüş',
 'Optik Sürücü',
 'Optik Sürücü Tipi',
 'Panel Tipi',
 'Price',
 'Ram (Sistem Belleği)',
 'Ram (Sistem Belleği) Tipi',
 'Renk',
 'SSD Kapasitesi',
 'Soket Tipi',
 'Star',
 'Temel İşlemci Hızı (GHz)',
 'Title',
 'Çözünürlük',
 'Çözünürlük Standartı',
 'İşlemci Frekansı',
 'İşlemci Modeli',
 'İşlemci Nesli',
 'İşlemci Tipi',
 'İşlemci Çekirdek Sayısı',
 'İşletim Sistemi',
 'Şarjlı Kullanım Süresi']

In [14]:
len(columns)

43

In [61]:
df = pd.DataFrame(columns = columns)

In [62]:
df

Empty DataFrame
Columns: [Arttırılabilir Azami Bellek, Bağlantılar, Brand, Cihaz Ağırlığı, Dokunmatik Ekran, Ekran Boyutu, Ekran Kartı, Ekran Kartı Bellek Tipi, Ekran Kartı Hafızası, Ekran Kartı Tipi, Ekran Yenileme Hızı, Ekran Özelliği, Garanti Süresi, Garanti Tipi, Görüntü Kalitesi, HDMI, Hard Disk Kapasitesi, Kapasite, Klavye, Kullanım Amacı, Maksimum İşlemci Hızı (GHz), Okula Dönüş, Optik Sürücü, Optik Sürücü Tipi, Panel Tipi, Price, Ram (Sistem Belleği), Ram (Sistem Belleği) Tipi, Renk, SSD Kapasitesi, Soket Tipi, Star, Temel İşlemci Hızı (GHz), Title, Çözünürlük, Çözünürlük Standartı, İşlemci Frekansı, İşlemci Modeli, İşlemci Nesli, İşlemci Tipi, İşlemci Çekirdek Sayısı, İşletim Sistemi, Şarjlı Kullanım Süresi]
Index: []

[0 rows x 43 columns]

In [37]:
computers

[[('Arttırılabilir Azami Bellek', 'NaN'),
  ('Bağlantılar', 'Bluetooth'),
  ('Brand',
   'Monster Abra A7 V13.2.3 Intel Core I5-11400h 16gb Ram 500gb Ssd 4gb Rtx3050ti Freedos 17.3" Fhd 144hz ABRA A7 V13.2.3 '),
  ('Cihaz Ağırlığı', '2 - 4 kg'),
  ('Dokunmatik Ekran', 'Yok'),
  ('Ekran Boyutu', '17,3 inç'),
  ('Ekran Kartı', 'Nvidia GeForce RTX 3050 Ti'),
  ('Ekran Kartı Bellek Tipi', 'GDDR6'),
  ('Ekran Kartı Hafızası', '4 GB'),
  ('Ekran Kartı Tipi', 'Harici'),
  ('Ekran Yenileme Hızı', '144 Hz'),
  ('Ekran Özelliği', 'NaN'),
  ('Garanti Süresi', 'NaN'),
  ('Garanti Tipi', 'Resmi Distribütör Garantili'),
  ('Görüntü Kalitesi', 'NaN'),
  ('HDMI', 'NaN'),
  ('Hard Disk Kapasitesi', 'Yok'),
  ('Kapasite', 'Yok'),
  ('Klavye', 'Q Türkçe'),
  ('Kullanım Amacı', 'Oyun'),
  ('Maksimum İşlemci Hızı (GHz)', '4.5'),
  ('Okula Dönüş', 'NaN'),
  ('Optik Sürücü', 'NaN'),
  ('Optik Sürücü Tipi', 'Yok'),
  ('Panel Tipi', 'IPS'),
  ('Price', '21.999'),
  ('Ram (Sistem Belleği)', '16 GB'),
  ('Ram (S

In [38]:
# her bilgisayar için eksik özellikleri NaN ile doldurma
for ix,computer in enumerate(computers):
    existing_attributes = set([attribute[0] for attribute in computer]) # var olan özellikler
    missing_attributes = set(df.columns) - existing_attributes
    for attribute in missing_attributes:
        computer.append((attribute,np.nan))
    computer.sort()

In [55]:
computer

[('Arttırılabilir Azami Bellek', nan),
 ('Bağlantılar', nan),
 ('Brand',
  'Dell G15 5511 G55112306u04 I5-11260h 64gb 512ssd Rtx3050 15.6" Fullhd Freedos Taşınabilir Bilgisayar T18781D '),
 ('Cihaz Ağırlığı', '2 - 4 kg'),
 ('Dokunmatik Ekran', 'Yok'),
 ('Ekran Boyutu', '15,6 inç'),
 ('Ekran Kartı', 'Nvidia GeForce RTX 3050'),
 ('Ekran Kartı Bellek Tipi', 'GDDR6'),
 ('Ekran Kartı Hafızası', '4 GB'),
 ('Ekran Kartı Tipi', 'Harici'),
 ('Ekran Yenileme Hızı', '120 Hz'),
 ('Ekran Özelliği', nan),
 ('Garanti Süresi', '2 Yıl'),
 ('Garanti Tipi', 'Resmi Distribütör Garantili'),
 ('Görüntü Kalitesi', nan),
 ('HDMI', nan),
 ('Hard Disk Kapasitesi', 'HDD Yok'),
 ('Kapasite', nan),
 ('Klavye', 'Q Türkçe'),
 ('Kullanım Amacı', 'Oyun'),
 ('Maksimum İşlemci Hızı (GHz)', nan),
 ('Okula Dönüş', nan),
 ('Optik Sürücü', nan),
 ('Optik Sürücü Tipi', nan),
 ('Panel Tipi', nan),
 ('Price', '29.999'),
 ('Ram (Sistem Belleği)', '64 GB'),
 ('Ram (Sistem Belleği) Tipi', 'DDR4'),
 ('Renk', 'Gri'),
 ('SSD Kapasit

In [56]:
computers_new =[]
for computer in computers:
    attributes_new = []
    for attribute in computer:
        attributes_new.append(attribute[1])
    computers_new.append(attributes_new)

In [57]:
computers_new

[[nan,
  'Bluetooth',
  'Monster Abra A7 V13.2.3 Intel Core I5-11400h 16gb Ram 500gb Ssd 4gb Rtx3050ti Freedos 17.3" Fhd 144hz ABRA A7 V13.2.3 ',
  '2 - 4 kg',
  'Yok',
  '17,3 inç',
  'Nvidia GeForce RTX 3050 Ti',
  'GDDR6',
  '4 GB',
  'Harici',
  '144 Hz',
  nan,
  nan,
  'Resmi Distribütör Garantili',
  nan,
  nan,
  'Yok',
  'Yok',
  'Q Türkçe',
  'Oyun',
  '4.5',
  nan,
  nan,
  'Yok',
  'IPS',
  '21.999',
  '16 GB',
  'DDR4',
  'Siyah',
  '500 GB',
  nan,
  4.64,
  nan,
  ' Abra A7 V13.2.3 Intel Core I5-11400h 16gb Ram 500gb Ssd 4gb Rtx3050ti Freedos 17.3" Fhd 144hz ABRA A7 V13.2.3 ',
  '1920 x 1080',
  'Full HD (FHD)',
  nan,
  '11400H',
  '11. Nesil',
  'Intel Core i5',
  '6',
  'Free Dos',
  '3 Saat ve Altı'],
 [nan,
  'Bluetooth',
  'Monster Abra A5 V16.7.3 Intel Core I5-11400h 16gb Ram 500gb Ssd 4gb Gtx1650 Freedos 15.6" Fhd 144hz ABRA A5 V16.7.3 ',
  '2 - 4 kg',
  'Yok',
  '15,6 inç',
  'Nvidia GeForce GTX 1650',
  'GDDR6',
  '4 GB',
  'Harici',
  '144 Hz',
  nan,
  '2 Yıl

In [63]:
for ix,content in enumerate(computers_new):
    df.loc[ix]= content

In [64]:
df

Arttırılabilir Azami Bellek Bağlantılar  \
0                            NaN   Bluetooth   
1                            NaN   Bluetooth   
2                            NaN   Bluetooth   
3                            NaN        HDMI   
4                           8 GB        HDMI   
...                          ...         ...   
4929                         NaN         NaN   
4930                         NaN         NaN   
4931                         NaN         NaN   
4932                         NaN         NaN   
4933                         NaN         NaN   

                                                  Brand Cihaz Ağırlığı  \
0     Monster Abra A7 V13.2.3 Intel Core I5-11400h 1...       2 - 4 kg   
1     Monster Abra A5 V16.7.3 Intel Core I5-11400h 1...       2 - 4 kg   
2     Monster Tulpar T7 V20.5.2 Intel Core I7-12700h...       2 - 4 kg   
3     Casper Excalibur G770.1265-dfj0x-b Intel Core ...       2 - 4 kg   
4     Apple Macbook Air 13'' M1 8gb 256gb Ssd Uzay G...  Belirtilmemiş   
...                                                 ...            ...   
4929  Dell G15 5511 G55112306u08 I5-11260h 64gb 512s...       2 - 4 kg   
4930  Dell G15 5511 G55112306u02 I5-11260h 32gb 512s...       2 - 4 kg   
4931  Dell G15 5511 G55112306u01 I5-11260h 16gb 1tbs...       2 - 4 kg   
4932  Dell G15 5511 G55112306u I5-11260h 16gb 512ssd...       2 - 4 kg   
4933  Dell G15 5511 G55112306u04 I5-11260h 64gb 512s...       2 - 4 kg   

     Dokunmatik Ekran Ekran Boyutu                 Ekran Kartı  \
0                 Yok     17,3 inç  Nvidia GeForce RTX 3050 Ti   
1                 Yok     15,6 inç     Nvidia GeForce GTX 1650   
2                 Yok     17,3 inç      Nvidia GeForce RTX3060   
3                 Yok     15,6 inç     Nvidia GeForce RTX 3050   
4                 Yok       13 inç             M1 7 Çekirdekli   
...               ...          ...                         ...   
4929              Yok     15,6 inç     Nvidia GeForce RTX 3050   
4930              Yok     15,6 inç     Nvidia GeForce RTX 3050   
4931              Yok     15,6 inç     Nvidia GeForce RTX 3050   
4932              Yok     15,6 inç     Nvidia GeForce RTX 3050   
4933              Yok     15,6 inç     Nvidia GeForce RTX 3050   

     Ekran Kartı Bellek Tipi Ekran Kartı Hafızası Ekran Kartı Tipi  ...  \
0                      GDDR6                 4 GB           Harici  ...   
1                      GDDR6                 4 GB           Harici  ...   
2                      GDDR6                 6 GB           Harici  ...   
3                      GDDR6                 4 GB           Harici  ...   
4              Belirtilmemiş           Paylaşımlı           Dahili  ...   
...                      ...                  ...              ...  ...   
4929                   GDDR6                 4 GB           Harici  ...   
4930                   GDDR6                 4 GB           Harici  ...   
4931                   GDDR6                 4 GB           Harici  ...   
4932                   GDDR6                 4 GB           Harici  ...   
4933                   GDDR6                 4 GB           Harici  ...   

                                                  Title   Çözünürlük  \
0      Abra A7 V13.2.3 Intel Core I5-11400h 16gb Ram...  1920 x 1080   
1      Abra A5 V16.7.3 Intel Core I5-11400h 16gb Ram...  1920 x 1080   
2      Tulpar T7 V20.5.2 Intel Core I7-12700h 16 Gb ...  1920 x 1080   
3      Excalibur G770.1265-dfj0x-b Intel Core I7-126...  1920 x 1080   
4      Macbook Air 13'' M1 8gb 256gb Ssd Uzay Grisi ...  2560 x 1600   
...                                                 ...          ...   
4929   G15 5511 G55112306u08 I5-11260h 64gb 512ssd R...  1920 x 1080   
4930   G15 5511 G55112306u02 I5-11260h 32gb 512ssd R...  1920 x 1080   
4931   G15 5511 G55112306u01 I5-11260h 16gb 1tbssd R...  1920 x 1080   
4932   G15 5511 G55112306u I5-11260h 16gb 512ssd Rtx...  1920 x 1080   
4933   G15 5511 G55112306u04 I5-11260h 64gb 512ssd R.

In [65]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4934 entries, 0 to 4933
Data columns (total 43 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Arttırılabilir Azami Bellek  326 non-null    object 
 1   Bağlantılar                  1457 non-null   object 
 2   Brand                        4934 non-null   object 
 3   Cihaz Ağırlığı               4934 non-null   object 
 4   Dokunmatik Ekran             4932 non-null   object 
 5   Ekran Boyutu                 4934 non-null   object 
 6   Ekran Kartı                  4934 non-null   object 
 7   Ekran Kartı Bellek Tipi      4932 non-null   object 
 8   Ekran Kartı Hafızası         4890 non-null   object 
 9   Ekran Kartı Tipi             4933 non-null   object 
 10  Ekran Yenileme Hızı          4666 non-null   object 
 11  Ekran Özelliği               154 non-null    object 
 12  Garanti Süresi               3950 non-null   object 
 13  Garanti Tipi      

In [66]:
df.to_csv('trendyol_laptops_4934.csv')

In [67]:
df.to_pickle("laptops_trendyol_4934") 